## Final Project A

<a id="item31"></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download the data and read it into a <em>pandas</em> dataframe.

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors

<a id="item1"></a>

<a id='item32'></a>

#### Let's go ahead and import the Keras library

In [11]:
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [12]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [13]:

# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

<a id="item4"></a>

<a id='item34'></a>

## Train and Test the Network

Let's call the function now to create our model.

In [14]:
# build the model
model = regression_model()

In [15]:
from sklearn.model_selection import train_test_split

## Notice I'm taking predictors which is NOT a Normalized Feature Set

In [16]:
X_train, X_test, Y_train, Y_test = train_test_split(predictors, target, test_size=0.3, random_state=200)

In [17]:
# fit the model
epochs = 50
model.fit(X_train, Y_train, epochs=50, verbose=1)

Epoch 1/50
721/721 [==============================] - 5s 7ms/step - loss: 2243.7715
Epoch 2/50
721/721 [==============================] - 2s 3ms/step - loss: 1234.7583: 0s -
Epoch 3/50
721/721 [==============================] - 2s 3ms/step - loss: 945.1783
Epoch 4/50
721/721 [==============================] - 2s 3ms/step - loss: 828.9105A: 0s - loss: 828.22
Epoch 5/50
721/721 [==============================] - 2s 2ms/step - loss: 737.7208
Epoch 6/50
721/721 [==============================] - 2s 2ms/step - loss: 662.4024
Epoch 7/50
721/721 [==============================] - 2s 2ms/step - loss: 599.6384
Epoch 8/50
721/721 [==============================] - 2s 3ms/step - loss: 546.0120A: 0s - loss
Epoch 9/50
721/721 [==============================] - 2s 3ms/step - loss: 501.3714
Epoch 10/50
721/721 [==============================] - 3s 4ms/step - loss: 464.0313
Epoch 11/50
721/721 [==============================] - 2s 2ms/step - loss: 431.0861
Epoch 12/50
 64/721 [=>......................

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.160261). Check your callbacks.
  % delta_t_median)


721/721 [==============================] - 2s 3ms/step - loss: 402.9678
Epoch 13/50
721/721 [==============================] - 3s 4ms/step - loss: 375.5404
Epoch 14/50
721/721 [==============================] - 2s 3ms/step - loss: 354.1318
Epoch 15/50
721/721 [==============================] - 3s 4ms/step - loss: 332.1399
Epoch 16/50
721/721 [==============================] - 2s 3ms/step - loss: 313.4217
Epoch 17/50
721/721 [==============================] - 2s 3ms/step - loss: 301.2433
Epoch 18/50
721/721 [==============================] - 2s 2ms/step - loss: 288.7821
Epoch 19/50
721/721 [==============================] - 3s 4ms/step - loss: 271.3080
Epoch 20/50
721/721 [==============================] - ETA: 0s - loss: 265.132 - 3s 5ms/step - loss: 263.3612
Epoch 21/50
721/721 [==============================] - 3s 4ms/step - loss: 251.7577
Epoch 22/50
721/721 [==============================] - 3s 4ms/step - loss: 234.3228A: 1s - 
Epoch 23/50
721/721 [==============================] -

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.140128). Check your callbacks.
  % delta_t_median)


721/721 [==============================] - 3s 4ms/step - loss: 208.4843
Epoch 26/50
721/721 [==============================] - 3s 4ms/step - loss: 198.5716
Epoch 27/50
721/721 [==============================] - 4s 5ms/step - loss: 191.8976
Epoch 28/50
721/721 [==============================] - 3s 5ms/step - loss: 187.6881
Epoch 29/50
721/721 [==============================] - 3s 4ms/step - loss: 178.5757
Epoch 30/50
721/721 [==============================] - 3s 4ms/step - loss: 172.2125
Epoch 31/50
721/721 [==============================] - 3s 4ms/step - loss: 165.1159
Epoch 32/50
721/721 [==============================] - 2s 3ms/step - loss: 160.6737
Epoch 33/50
721/721 [==============================] - 3s 4ms/step - loss: 156.6541
Epoch 34/50
721/721 [==============================] - 3s 5ms/step - loss: 151.0941
Epoch 35/50
721/721 [==============================] - 3s 4ms/step - loss: 147.1789
Epoch 36/50
721/721 [==============================] - 2s 3ms/step - loss: 144.1267
Epoc

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.102354). Check your callbacks.
  % delta_t_median)


721/721 [==============================] - 3s 4ms/step - loss: 126.5796
Epoch 45/50
721/721 [==============================] - 3s 4ms/step - loss: 126.3849A: 1s 
Epoch 46/50
721/721 [==============================] - 3s 4ms/step - loss: 122.7702
Epoch 47/50
721/721 [==============================] - 2s 3ms/step - loss: 121.4777
Epoch 48/50
721/721 [==============================] - 2s 3ms/step - loss: 120.1229
Epoch 49/50
 64/721 [=>............................] - ETA: 1s - loss: 132.1850

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.100624). Check your callbacks.
  % delta_t_median)


721/721 [==============================] - 3s 5ms/step - loss: 118.8530
Epoch 50/50
721/721 [==============================] - 3s 4ms/step - loss: 118.2628


In [18]:
loss_val = model.evaluate(X_test, Y_test)
Y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 1s 4ms/step


100.44453319068094

In [19]:
#Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.
from sklearn.metrics import mean_squared_error

In [20]:
mse = mean_squared_error(Y_test, Y_pred)
mean = np.mean(mse)
standard_deviation = np.std(mse)
print(mean, standard_deviation)

100.44453393859304 0.0


In [24]:
total_mses = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mses):
    X_train, X_test, Y_train, Y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, Y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, Y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    Y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(Y_test, Y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mses) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 53.098714871699755
MSE 2: 65.57577978831665
MSE 3: 49.692688296913715
MSE 4: 54.949095790826
MSE 5: 53.85659230797036
MSE 6: 54.415949824558496
MSE 7: 60.12421083913266
MSE 8: 46.14864555531721
MSE 9: 47.571929042779125
MSE 10: 51.772936935177896
MSE 11: 46.3023163631896
MSE 12: 44.70026094010733
MSE 13: 51.09525181174664
MSE 14: 53.13979295156535
MSE 15: 45.52974854009437
MSE 16: 40.549344257243625
MSE 17: 45.85358704255237
MSE 18: 45.87535998890701
MSE 19: 45.43633848023646
MSE 20: 46.78755321317506
MSE 21: 46.61251359464281
MSE 22: 44.76602683946924
MSE 23: 41.09928207644367
MSE 24: 48.33019775094338
MSE 25: 50.23619632659221
MSE 26: 47.7303019156348
MSE 27: 44.84707292229612
MSE 28: 42.17175491728057
MSE 29: 52.49442328295662
MSE 30: 49.06498981833844
MSE 31: 47.22562931887926
MSE 32: 40.20846777980768
MSE 33: 42.49134827894686
MSE 34: 45.41215759573631
MSE 35: 45.9414536559466
MSE 36: 51.20593020985427
MSE 37: 48.60436339517241
MSE 38: 51.07970601300977
MSE 39: 48.297349145